csp/adapters/arrow.py

```diff
@@ -0,0 +1,253 @@
+                try:
+                    start_idx, next_start_idx = next(self.chunk_index_iter)
+                    new_batches = [self.batch.slice(offset=start_idx, length=next_start_idx - start_idx)]
+                    new_ts = self.batch[self.ts_col_name][start_idx].as_py()
+                    self.processed_chunks_count += 1
+                    if self.last_chunk:
+                        if self.last_ts == new_ts:
+                            new_batches = self.last_chunk + new_batches
+                            self.last_chunk = None
+                            self.last_ts = None
+                        else:
+                            raise Exception("last_chunk and new_batches have different timestamps")
+
+                    if self.processed_chunks_count == self.batch_chunks_count:
+                        self.last_chunk = new_batches
+                        self.last_ts = new_ts
+                        self.processed_chunks_count = 0
+                    else:
+                        if new_ts > self.end_time:
+                            self.finished = True
+                            continue
+                        return (new_ts, new_batches)
+                except StopIteration:
+                    raise Exception("chunk_index_iter reached end, how?")
+
+            # Try to get a new batch of data
+            try:
+                if self.starting_batch:
+                    # Use the sliced batch from start_time filtering
+                    self.batch = self.starting_batch
+                    self.starting_batch = None
+                else:
+                    # Get the next batch of data
+                    self.batch = next(self.source)
+                    if self.batch.num_rows == 0:
+                        continue
+
+                all_timestamps = self.batch[self.ts_col_name]
+                unique_timestamps = all_timestamps.unique()
+                indexes = pc.index_in(unique_timestamps, all_timestamps).to_pylist() + [self.batch.num_rows]
+                self.chunk_index_iter = zip(indexes, indexes[1:])
+                self.batch_chunks_count = len(unique_timestamps)
+                starting_ts = unique_timestamps[0].as_py()
+                if starting_ts != self.last_ts and self.last_chunk:
+                    new_batches = self.last_chunk
+                    new_ts = self.last_ts
+                    self.last_chunk = None
+                    self.last_ts = None
+                    if new_ts > self.end_time:
+                        self.finished = True
+                        continue
+                    return (new_ts, new_batches)
+            except StopIteration:
+                self.finished = True
+                if self.last_chunk:
+                    if self.last_ts > self.end_time:
+                        continue
+                    return (self.last_ts, self.last_chunk)
+        return None
+
+
+ArrowHistoricalAdapter = py_pull_adapter_def(
```

Comment:
Does the docstring for this adapter automatically pull the docstring for the adapter impl? We should make sure it has good documentation